### PR TITLE
Update SSO to match Agent Connect requirements

### PR DESCRIPTION
**Pull request by lebaudantoine**
## Purpose

Prepare Desk deployment which will run the Agent Connect SSO.

## Proposal

Replace the Implicit flow by a standard and secure Authorization Code flow.
Agent Connect is the official SSO, and requires it. [Here is their documentation](https://github.com/france-connect/Documentation-AgentConnect/tree/main).

For a better developer experience, we'll stick to a Keycloack server managing our OIDC flow in development. Please feel free to ask help to configure Agent Connect Integration environment with our app. I am using Ngrok with a static IP.

Note: The OIDC spec recommends adding PKCE to the Authorization Code flow when working with browser based apps / SPA. This security mechanism is a strong recommendation, but not a requirement. It's not yet released by Agent Connect. They panned to ship it by the end of June.

🚧 some commits will be consolidated into a functional commit upon PR merging ! 

[A small demo 👀](https://www.loom.com/share/14b6db9a651a4e4fbdbca105fd351283)

----
I still need to fix e2e tests, locally and in the CI.

### CODE REVIEW

File: src/backend/core/authentication.py [hunk 1]
```diff
@@ -1,59 +1,111 @@
...
-    def get_security_definition(self, auto_schema):
-        """Return the security definition for JWT authentication."""
-        return build_bearer_security_scheme_object(
-            header_name="Authorization",
-            token_prefix="Bearer",  # noqa S106
+        user = (
+            self.UserModel.objects.filter(identities__sub=sub)
+            .annotate(identity_email=models.F("identities__email"))
+            .distinct()
+            .first()
```
Comment (lebaudantoine) [inline src/backend/core/authentication.py:85]: Discussed IRL, this logic will be updated in a dedicated PR.

File: src/frontend/apps/desk/src/api/__tests__/fetchApi.test.tsx [hunk 2]
```diff
@@ -20,29 +19,25 @@ describe('fetchAPI', () => {
...
 
   it('logout if 401 response', async () => {
-    useAuthStore.setState({ token: 'my-token' });
-
     fetchMock.mock('http://some.api.url/api/v1.0/some/url', 401);
 
     await fetchAPI('some/url');
 
-    expect(useAuthStore.getState().token).toBeNull();
+    expect(useAuthStore.getState().userData).toBeUndefined();
```
Comment (AntoLC) [inline src/frontend/apps/desk/src/api/__tests__/fetchApi.test.tsx:41]: How do you know if it was not `undefined` from the beginning of the test ?

File: src/frontend/apps/desk/src/features/auth/api/getMe.tsx [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+ * This interface is incomplete, and will be
+ * refactored in a near future.
+ *
+ * @interface UserData
+ * @property {string} email - The email of the user.
+ */
+export interface UserData {
+  email: string;
+}
+
```
Comment (AntoLC) [inline src/frontend/apps/desk/src/features/auth/api/getMe.tsx:14]: Why not interfaced now? you have the values returned already.

File: src/frontend/apps/desk/src/features/auth/useAuthStore.tsx [hunk 4]
```diff
@@ -1,40 +1,47 @@
...
+        return {};
       }
-
-      /**
-       * TODO: Implement OIDC production authentication
-       */
-
+      getMe()
+        .then((data: UserData) => {
+          set({ initialized: true, authenticated: true, userData: data });
```
Comment (AntoLC) [inline src/frontend/apps/desk/src/features/auth/useAuthStore.tsx:31]: As it is now, `initialized` and `authenticated` are doing the same job.

File: src/frontend/apps/desk/src/features/auth/useAuthStore.tsx [hunk 5]
```diff
@@ -1,40 +1,47 @@
...
+          return { initialized: true };
+        })
+        .catch(() => {
+          // todo - implement a proper login screen to prevent automatic navigation.
```
Comment (AntoLC) [inline src/frontend/apps/desk/src/features/auth/useAuthStore.tsx:35]: FMI, Agent connect doesn't manage the login screen ?

File: src/frontend/apps/desk/.env.development [hunk 6]
```diff
@@ -1,5 +1 @@
 NEXT_PUBLIC_API_URL=http://localhost:8071/api/v1.0/
```
Comment (AntoLC) [inline src/frontend/apps/desk/.env.development:1]: You can clean them here as well so:
https://github.com/numerique-gouv/people/blob/main/src/frontend/apps/desk/src/custom-next.d.ts#L23-L26

File: docker-compose.yml [hunk 7]
```diff
@@ -161,6 +168,7 @@ services:
       KC_DB_PASSWORD: pass
       KC_DB_USERNAME: people
       KC_DB_SCHEMA: public
+      PROXY_ADDRESS_FORWARDING: true
```
Comment (AntoLC) [inline docker-compose.yml:171]: The job `e2e` doesn't seem happy with that. Considere maybe a string ?

File: src/frontend/apps/desk/src/features/auth/useAuthStore.tsx
Comment (AntoLC) [inline src/frontend/apps/desk/src/features/auth/useAuthStore.tsx:1]: Small state issue I think.
I don't know if it is as it should be about the 301.
image

Comment (AntoLC) [inline src/frontend/apps/desk/src/features/auth/useAuthStore.tsx:1]: I get a 403 when I create a team:
`detail: "CSRF Failed: Origin checking failed - http://localhost:3000 does not match any trusted origins."`
image

**Review by AntoLC — COMMENTED** [review]
Some things to check but nice job, I guess it is not easy to cohabit AgentConnect with Keycloak. Nice commits btw 😉 
I mainly focused on the front part, Sam will be more skills than me about this part.

File: src/frontend/apps/desk/src/features/auth/useAuthStore.tsx (hunk 5 above)
Comment (lebaudantoine) [reply src/frontend/apps/desk/src/features/auth/useAuthStore.tsx:35]: Sorry, I mean the first screen on the Figma. There is a landing page with the Agent Connect button at the top right. I can update my comment.

File: src/frontend/apps/desk/src/features/auth/api/getMe.tsx (hunk 3 above)
Comment (lebaudantoine) [reply src/frontend/apps/desk/src/features/auth/api/getMe.tsx:14]: Interface will need to be updated with first name and last name right ?

File: docker-compose.yml (hunk 7 above)
Comment (lebaudantoine) [reply docker-compose.yml:171]: Yeah I wanted to discuss it, I followed your initial implementation. What type of string? 
Btw, e2e tests need to be fixed locally !

File: src/frontend/apps/desk/.env.development (hunk 6 above)
Comment (lebaudantoine) [reply src/frontend/apps/desk/.env.development:1]: done !

File: src/frontend/apps/desk/src/features/auth/useAuthStore.tsx (hunk 5 above)
Comment (lebaudantoine) [reply src/frontend/apps/desk/src/features/auth/useAuthStore.tsx:35]: Capture d’écran 2024-02-16 à 11 05 42

File: src/backend/core/authentication.py [hunk 8]
```diff
@@ -1,59 +1,111 @@
...
-from rest_framework_simplejwt.authentication import JWTAuthentication
+import requests
+from mozilla_django_oidc.auth import (
+    OIDCAuthenticationBackend as MozillaOIDCAuthenticationBackend,
+)
+from rest_framework_simplejwt.exceptions import InvalidToken
+
+from .models import Identity
+
+LOGGER = logging.getLogger(__name__)
```
[the remark this answers is not in the document]
Comment (lebaudantoine) [reply src/backend/core/authentication.py]: weird, it passed the back-end linter. Idk if unused constant are ignored, or logger can be unused. Fixed !

File: src/frontend/apps/desk/src/features/auth/useAuthStore.tsx
Comment (lebaudantoine) [reply src/frontend/apps/desk/src/features/auth/useAuthStore.tsx:1]: 301, are because some redirections happened for the Login flow.
When there is no redirection, it returns a 200.

Comment (lebaudantoine) [reply src/frontend/apps/desk/src/features/auth/useAuthStore.tsx:1]: Actually, it not a "state" issue. React is running in strict mode, it triggers twice side effects in development. This doesn't cause any issue, and won't run twice in production. It highlights that we might have a design issue in our code. Handling this login logic in an Effect.

Comment (lebaudantoine) [reply src/frontend/apps/desk/src/features/auth/useAuthStore.tsx:1]: I plan to refactor properly this piece of logic, using `react-querry` when implementing the homepage. Using `useQuery` will simplify and solve this problem.

File: src/frontend/apps/desk/src/features/auth/useAuthStore.tsx (hunk 4 above)
Comment (lebaudantoine) [reply src/frontend/apps/desk/src/features/auth/useAuthStore.tsx]: I reverted it to `set({ authenticated: true, userData: data });` is that okay?

File: src/frontend/apps/desk/src/api/__tests__/fetchApi.test.tsx (hunk 2 above)
Comment (lebaudantoine) [reply src/frontend/apps/desk/src/api/__tests__/fetchApi.test.tsx:43]: fixed !

File: src/frontend/apps/desk/src/features/auth/useAuthStore.tsx
Comment (lebaudantoine) [reply src/frontend/apps/desk/src/features/auth/useAuthStore.tsx:1]: Sorry I forgot to test it, my bad !! I fixed it by adding the right header to the request.

File: docker-compose.yml (hunk 7 above)
Comment (lebaudantoine) [reply docker-compose.yml:171]: Happy to dig it together !

File: src/backend/pyproject.toml [hunk 9]
```diff
@@ -50,6 +50,7 @@ dependencies = [
     "sentry-sdk==1.40.3",
     "url-normalize==1.4.3",
     "whitenoise==6.6.0",
+    "mozilla-django-oidc==4.0.0",
```
Comment (sampaccoud) [inline src/backend/pyproject.toml:53]: remove "djangorestframework-simplejwt[crypto]==5.3.1", ?

File: .github/workflows/people.yml [hunk 10]
```diff
@@ -216,7 +216,7 @@ jobs:
       DJANGO_CONFIGURATION: Test
       DJANGO_SETTINGS_MODULE: people.settings
       DJANGO_SECRET_KEY: ThisIsAnExampleKeyForTestPurposeOnly
-      DJANGO_JWT_PRIVATE_SIGNING_KEY: ThisIsAnExampleKeyForDevPurposeOnly
```
Comment (sampaccoud) [inline .github/workflows/people.yml:219]: I think you should open an issue with the 3 TODO listed in the commit

File: docker-compose.yml (hunk 7 above)
Comment (AntoLC) [reply docker-compose.yml:171]: https://docs.docker.com/compose/compose-file/compose-file-v3/#environment
So I guess:
```suggestion
      PROXY_ADDRESS_FORWARDING: 'true'
```

Let's see what the job e2e says, it will be easier to debug in local after that.

File: src/frontend/apps/desk/src/features/auth/useAuthStore.tsx [hunk 11]
```diff
@@ -1,40 +1,47 @@
...
+          window.location.replace(
+            new URL('authenticate', process.env.NEXT_PUBLIC_API_URL).href,
+          );
+        });
       return {};
     }),
```
Comment (AntoLC) [inline src/frontend/apps/desk/src/features/auth/useAuthStore.tsx:41]: About this conversation (https://github.com/numerique-gouv/people/pull/59#discussion_r1492177860)

It will fix the duplicate call, getMe is async:
```suggestion
      getMe()
        .then((data: UserData) => {
          set({ authenticated: true, userData: data });
        })
        .catch(() => {
          // todo - implement a proper login screen to prevent automatic navigation.
          window.location.replace(
            new URL('authenticate', process.env.NEXT_PUBLIC_API_URL).href,
          );
        });
      return { initialized: true };
```

File: src/frontend/apps/desk/src/features/auth/useAuthStore.tsx
Comment (AntoLC) [reply src/frontend/apps/desk/src/features/auth/useAuthStore.tsx:1]: Working, top.

File: src/frontend/apps/desk/src/features/auth/useAuthStore.tsx (hunk 11 above)
Comment (lebaudantoine) [reply src/frontend/apps/desk/src/features/auth/useAuthStore.tsx:41]: I will apply your suggestion !

File: src/frontend/apps/desk/src/api/fetchApi.ts [hunk 12]
```diff
@@ -1,18 +1,39 @@
...
+  const cookies = document.cookie.split(';');
+  for (let i = 0; i < cookies.length; i++) {
+    const cookie = cookies[i].trim();
+    if (cookie.startsWith('csrftoken=')) {
+      return cookie.substring('csrftoken='.length, cookie.length);
+    }
+  }
+  return null;
+}
+
```
[the remark this answers is not in the document]
Comment (lebaudantoine) [reply src/frontend/apps/desk/src/api/fetchApi.ts:18]: fixed !

File: src/frontend/apps/desk/src/features/auth/useAuthStore.tsx (hunk 11 above)
Comment (lebaudantoine) [reply src/frontend/apps/desk/src/features/auth/useAuthStore.tsx]: Actually, this suggestion is not working. It triggers dozens of `getMe` request, while mine initial code was only triggering request twice due to the React strict mode.

File: src/frontend/apps/desk/src/features/auth/Auth.tsx [hunk 13]
```diff
@@ -6,15 +6,11 @@ import { Box } from '@/components';
...
-  const { initAuth, authenticated, initialized } = useAuthStore();
+  const { authenticated, initAuth } = useAuthStore();
 
   useEffect(() => {
-    if (initialized) {
-      return;
-    }
-
     initAuth();
-  }, [initAuth, initialized]);
```
[the remark this answers is not in the document]
Comment (AntoLC) [reply src/frontend/apps/desk/src/features/auth/Auth.tsx:17]: [scrnli_2_21_2024_5-04-08 PM.webm](https://github.com/numerique-gouv/people/assets/25994652/ab55f913-5744-40b6-abb4-ea79f57a33db)

[the remark this answers is not in the document]
Comment (lebaudantoine) [reply src/frontend/apps/desk/src/features/auth/Auth.tsx:17]: I have fixed it ! by redirecting to the login screen when detecting a 401. My bad.

File: src/backend/core/authentication.py [hunk 14]
```diff
@@ -1,59 +1,107 @@
...
-from rest_framework_simplejwt.authentication import JWTAuthentication
+import requests
+from mozilla_django_oidc.auth import (
+    OIDCAuthenticationBackend as MozillaOIDCAuthenticationBackend,
+)
+
+from .models import Identity
+
 
+class OIDCAuthenticationBackend(MozillaOIDCAuthenticationBackend):
```
Comment (sampaccoud) [inline src/backend/core/authentication.py:15]: Since you are mentioning customizations for AgentConnect, should we name it AgenceConnectOIDCAuthenticationBackend? Or is it in fact generic and we can remove the mention to AgentConnect in the docstring?

**Review by sampaccoud — APPROVED** [review]

File: src/frontend/apps/desk/src/api/fetchApi.ts [hunk 15]
```diff
@@ -31,7 +31,12 @@ export const fetchAPI = async (input: string, init?: RequestInit) => {
 
   // todo - handle 401, redirect to login screen
   // todo - please have a look to this documentation page https://mozilla-django-oidc.readthedocs.io/en/stable/xhr.html
-  response.status === 401 && logout();
+  if (response.status === 401) {
+    logout();
+    window.location.replace(
+      new URL('authenticate/', process.env.NEXT_PUBLIC_API_URL).href,
+    );
```
Comment (AntoLC) [inline src/frontend/apps/desk/src/api/fetchApi.ts:38]: Could we keep the logic of the authentication in the auth feature ? The function `logout` should do the redirection so, let's keep it DRY moreover,  we have this part in 2 different places, not easy for a future refacto.

Comment (lebaudantoine) [reply src/frontend/apps/desk/src/api/fetchApi.ts:38]: As I mentioned, I will handle this refacto, moreover, this code doesn't log out the user, but re-login her. Can totally encapsulate the navigation in the Auth feature.

**Review by AntoLC — APPROVED** [review]
Good thing done ✅
